diff -ruN original_dirview/CMakeLists.txt modified_dirview/CMakeLists.txt
--- original_dirview/CMakeLists.txt	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/CMakeLists.txt	2026-06-28 23:26:28.457587170 +0300
@@ -0,0 +1,49 @@
+cmake_minimum_required(VERSION 3.5)
+
+project(qt-file-system-explorer VERSION 0.1 LANGUAGES CXX)
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+set(CMAKE_AUTOUIC ON)
+set(CMAKE_AUTOMOC ON)
+set(CMAKE_AUTORCC ON)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+find_package(QT NAMES Qt6 Qt5 REQUIRED COMPONENTS Widgets)
+find_package(Qt${QT_VERSION_MAJOR} REQUIRED COMPONENTS Core Gui Widgets)
+
+add_subdirectory(utils)
+
+option(BUILD_TESTS "Build unit tests" OFF)
+
+set(PROJECT_SOURCES
+        src/main.cpp
+		src/main_window.h
+		src/main_window.cpp
+		src/ui_parts/custom_system_model_delegate.h
+		src/ui_parts/custom_system_model_delegate.cpp
+		src/ui_parts/custom_system_model.h
+		src/ui_parts/custom_system_model.cpp
+		src/ui_parts/custom_sort_filter_proxy_model.h
+		src/ui_parts/custom_sort_filter_proxy_model.cpp
+)
+
+add_executable(${PROJECT_NAME}
+	${PROJECT_SOURCES}
+)
+
+target_link_libraries(${PROJECT_NAME} PRIVATE
+	utils
+	Qt${QT_VERSION_MAJOR}::Widgets
+)
+
+install(TARGETS ${PROJECT_NAME}
+	RUNTIME DESTINATION bin
+)
+
+if(BUILD_TESTS)
+	enable_testing()
+	add_subdirectory(tests)
+endif()
diff -ruN original_dirview/dirview.pro modified_dirview/dirview.pro
--- original_dirview/dirview.pro	2026-06-20 21:30:57.221197787 +0300
+++ modified_dirview/dirview.pro	1970-01-01 03:00:00.000000000 +0300
@@ -1,8 +0,0 @@
-QT += widgets
-requires(qtConfig(treeview))
-
-SOURCES       = main.cpp
-
-# install
-target.path = $$[QT_INSTALL_EXAMPLES]/widgets/itemviews/dirview
-INSTALLS += target
diff -ruN original_dirview/main.cpp modified_dirview/main.cpp
--- original_dirview/main.cpp	2026-06-20 21:30:57.221197787 +0300
+++ modified_dirview/main.cpp	1970-01-01 03:00:00.000000000 +0300
@@ -1,107 +0,0 @@
-/****************************************************************************
-**
-** Copyright (C) 2020 The Qt Company Ltd.
-** Contact: https://www.qt.io/licensing/
-**
-** This file is part of the examples of the Qt Toolkit.
-**
-** $QT_BEGIN_LICENSE:BSD$
-** Commercial License Usage
-** Licensees holding valid commercial Qt licenses may use this file in
-** accordance with the commercial license agreement provided with the
-** Software or, alternatively, in accordance with the terms contained in
-** a written agreement between you and The Qt Company. For licensing terms
-** and conditions see https://www.qt.io/terms-conditions. For further
-** information use the contact form at https://www.qt.io/contact-us.
-**
-** BSD License Usage
-** Alternatively, you may use this file under the terms of the BSD license
-** as follows:
-**
-** "Redistribution and use in source and binary forms, with or without
-** modification, are permitted provided that the following conditions are
-** met:
-**   * Redistributions of source code must retain the above copyright
-**     notice, this list of conditions and the following disclaimer.
-**   * Redistributions in binary form must reproduce the above copyright
-**     notice, this list of conditions and the following disclaimer in
-**     the documentation and/or other materials provided with the
-**     distribution.
-**   * Neither the name of The Qt Company Ltd nor the names of its
-**     contributors may be used to endorse or promote products derived
-**     from this software without specific prior written permission.
-**
-**
-** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
-** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
-** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
-** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
-** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
-** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
-** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
-** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
-**
-** $QT_END_LICENSE$
-**
-****************************************************************************/
-
-#include <QApplication>
-#include <QFileSystemModel>
-#include <QFileIconProvider>
-#include <QScreen>
-#include <QScroller>
-#include <QTreeView>
-#include <QCommandLineParser>
-#include <QCommandLineOption>
-
-int main(int argc, char *argv[])
-{
-    QApplication app(argc, argv);
-
-    QCoreApplication::setApplicationVersion(QT_VERSION_STR);
-    QCommandLineParser parser;
-    parser.setApplicationDescription("Qt Dir View Example");
-    parser.addHelpOption();
-    parser.addVersionOption();
-    QCommandLineOption dontUseCustomDirectoryIconsOption("c", "Set QFileSystemModel::DontUseCustomDirectoryIcons");
-    parser.addOption(dontUseCustomDirectoryIconsOption);
-    QCommandLineOption dontWatchOption("w", "Set QFileSystemModel::DontWatch");
-    parser.addOption(dontWatchOption);
-    parser.addPositionalArgument("directory", "The directory to start in.");
-    parser.process(app);
-    const QString rootPath = parser.positionalArguments().isEmpty()
-        ? QString() : parser.positionalArguments().first();
-
-    QFileSystemModel model;
-    model.setRootPath("");
-    if (parser.isSet(dontUseCustomDirectoryIconsOption))
-        model.setOption(QFileSystemModel::DontUseCustomDirectoryIcons);
-    if (parser.isSet(dontWatchOption))
-        model.setOption(QFileSystemModel::DontWatchForChanges);
-    QTreeView tree;
-    tree.setModel(&model);
-    if (!rootPath.isEmpty()) {
-        const QModelIndex rootIndex = model.index(QDir::cleanPath(rootPath));
-        if (rootIndex.isValid())
-            tree.setRootIndex(rootIndex);
-    }
-
-    // Demonstrating look and feel features
-    tree.setAnimated(false);
-    tree.setIndentation(20);
-    tree.setSortingEnabled(true);
-    const QSize availableSize = tree.screen()->availableGeometry().size();
-    tree.resize(availableSize / 2);
-    tree.setColumnWidth(0, tree.width() / 3);
-
-    // Make it flickable on touchscreens
-    QScroller::grabGesture(&tree, QScroller::TouchGesture);
-
-    tree.setWindowTitle(QObject::tr("Dir View"));
-    tree.show();
-
-    return app.exec();
-}
diff -ruN original_dirview/src/main.cpp modified_dirview/src/main.cpp
--- original_dirview/src/main.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/main.cpp	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,12 @@
+#include <QApplication>
+#include "main_window.h"
+
+int main(int argc, char *argv[])
+{
+	QApplication a(argc, argv);
+
+	MainWindow w;
+	w.show();
+
+	return a.exec();
+}
diff -ruN original_dirview/src/main_window.cpp modified_dirview/src/main_window.cpp
--- original_dirview/src/main_window.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/main_window.cpp	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,126 @@
+#include <QDir>
+#include <QThread>
+#include <QGuiApplication>
+#include <QScreen>
+
+#include "main_window.h"
+
+MainWindow::MainWindow(QWidget *parent)
+	: QMainWindow(parent),
+	  m_layout{new QVBoxLayout},
+	  m_treeView{new QTreeView},
+	  m_lineEdit{new QLineEdit},
+	  m_thPool{new QThreadPool(this)},
+	  m_proxyModel{new CustomSortFilterProxyModel(this)},
+	  m_delegate{new CustomSystemModelDelegate(this)},
+	  m_fsModel{new CustomSystemModel(m_thPool, &m_stopFlag, this)},
+	  m_rootPath{QDir::homePath()}
+{
+	int maxThreadCount = QThread::idealThreadCount();
+
+	if (maxThreadCount < 2)
+	{
+		maxThreadCount = 2;
+	}
+
+	m_thPool->setMaxThreadCount(
+		qMax(1, static_cast<int>(maxThreadCount * kThreadPoolUsageRatio))
+	);
+
+	m_fsModel->setFilter(
+		QDir::AllEntries |
+		QDir::Hidden |
+		QDir::NoDotAndDotDot |
+		QDir::NoSymLinks
+	);
+	m_fsModel->setRootPath(m_rootPath);
+
+	m_proxyModel->setSourceModel(m_fsModel);
+	m_proxyModel->setFilterKeyColumn(0);
+	m_proxyModel->setFilterCaseSensitivity(Qt::CaseInsensitive);
+	m_proxyModel->setRecursiveFilteringEnabled(true);
+
+	m_setupUi();
+
+	connect(
+		m_treeView,
+		&QTreeView::clicked,
+		this,
+		&MainWindow::m_onTreeViewItemClicked
+	);
+	connect(
+		m_lineEdit,
+		&QLineEdit::textChanged,
+		this,
+		&MainWindow::m_onLineEditTextChanged
+	);
+}
+
+MainWindow::~MainWindow()
+{
+	m_stopFlag.setFlag(true);
+	m_thPool->clear();
+	m_thPool->waitForDone();
+}
+
+void MainWindow::m_setupUi()
+{
+	auto* widget = new QWidget(this);
+
+	m_layout->addWidget(m_lineEdit);
+	m_layout->addWidget(m_treeView);
+	widget->setLayout(m_layout);
+
+	m_treeView->setModel(m_proxyModel);
+	m_treeView->setItemDelegateForColumn(1, m_delegate);
+	m_treeView->setAnimated(false);
+	m_treeView->setIndentation(kTreeViewIndentation);
+	m_treeView->setSortingEnabled(true);
+	m_resetRootIndex();
+
+
+	setCentralWidget(widget);
+	setWindowTitle(kWindowTitle);
+
+	// Centering the window.
+	int x = 0, y = 0;
+
+	if (QScreen* screen = QGuiApplication::primaryScreen())
+	{
+		const QRect screenGeometry = screen->availableGeometry();
+		x = (screenGeometry.width() - kDefaultWindowWidth) / 2;
+		y = (screenGeometry.height() - kDefaultWindowHeight) / 2;
+	}
+
+	setGeometry(x, y, kDefaultWindowWidth, kDefaultWindowHeight);
+}
+
+void MainWindow::m_onLineEditTextChanged(const QString& text)
+{
+	m_proxyModel->setFilterWildcard(
+		QString("*%1*").arg(text)
+	);
+	m_resetRootIndex();
+}
+
+void MainWindow::m_onTreeViewItemClicked(const QModelIndex& index)
+{
+	if (!index.isValid() || index.column() != 1)
+	{
+		return;
+	}
+
+	if (index.data(Qt::DisplayRole).toString().isEmpty())
+	{
+		m_fsModel->calculateAndSetDirSize(m_proxyModel->mapToSource(index));
+	}
+}
+
+void MainWindow::m_resetRootIndex()
+{
+	m_treeView->setRootIndex(
+		m_proxyModel->mapFromSource(
+			m_fsModel->index(m_rootPath)
+		)
+	);
+}
diff -ruN original_dirview/src/main_window.h modified_dirview/src/main_window.h
--- original_dirview/src/main_window.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/main_window.h	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,48 @@
+#ifndef MAIN_WINDOW_H
+#define MAIN_WINDOW_H
+
+#include <QMainWindow>
+#include <QTreeView>
+#include <QVBoxLayout>
+#include <QLineEdit>
+#include <QThreadPool>
+#include <QModelIndex>
+
+#include "utils/file_utils.h"
+#include "utils/types.h"
+
+#include "src/ui_parts/custom_system_model.h"
+#include "src/ui_parts/custom_sort_filter_proxy_model.h"
+#include "src/ui_parts/custom_system_model_delegate.h"
+
+class MainWindow : public QMainWindow
+{
+	Q_OBJECT
+public:
+	explicit MainWindow(QWidget *parent = nullptr);
+	~MainWindow();
+private slots:
+	void m_onTreeViewItemClicked(const QModelIndex& index);
+	void m_onLineEditTextChanged(const QString& text);
+private:
+	QVBoxLayout* m_layout;
+	QTreeView* m_treeView;
+	QLineEdit* m_lineEdit;
+	QThreadPool* m_thPool;
+	StopFlag m_stopFlag;
+	CustomSortFilterProxyModel* m_proxyModel;
+	CustomSystemModelDelegate* m_delegate;
+	CustomSystemModel* m_fsModel;
+	const QString m_rootPath;
+
+	void m_setupUi();
+	void m_resetRootIndex();
+
+	static constexpr int kDefaultWindowWidth = 800;
+	static constexpr int kDefaultWindowHeight = 400;
+	static constexpr int kTreeViewIndentation = 20;
+	static constexpr double kThreadPoolUsageRatio = 0.5;
+	static constexpr const char* kWindowTitle = "dirview with dir size calculation";
+};
+
+#endif // MAIN_WINDOW_H
diff -ruN original_dirview/src/ui_parts/custom_sort_filter_proxy_model.cpp modified_dirview/src/ui_parts/custom_sort_filter_proxy_model.cpp
--- original_dirview/src/ui_parts/custom_sort_filter_proxy_model.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_sort_filter_proxy_model.cpp	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,32 @@
+#include "custom_sort_filter_proxy_model.h"
+#include "utils/types.h"
+
+CustomSortFilterProxyModel::CustomSortFilterProxyModel(QObject* parent)
+	: QSortFilterProxyModel(parent)
+{}
+
+bool CustomSortFilterProxyModel::filterAcceptsRow(int source_row, const QModelIndex& source_parent) const
+{
+	// Guard against an unrepresented index.
+	if (!source_parent.isValid())
+	{
+		return false;
+	}
+
+	return QSortFilterProxyModel::filterAcceptsRow(source_row, source_parent);
+}
+
+bool CustomSortFilterProxyModel::lessThan(const QModelIndex &source_left, const QModelIndex &source_right) const
+{
+	int column = source_left.column();
+
+	if (column == 1)
+	{
+		auto leftSize = sourceModel()->data(source_left, CustomRoles::SizeInBytesRole).toLongLong();
+		auto rightSize = sourceModel()->data(source_right, CustomRoles::SizeInBytesRole).toLongLong();
+
+		return leftSize < rightSize;
+	}
+
+	return QSortFilterProxyModel::lessThan(source_left, source_right);
+}
diff -ruN original_dirview/src/ui_parts/custom_sort_filter_proxy_model.h modified_dirview/src/ui_parts/custom_sort_filter_proxy_model.h
--- original_dirview/src/ui_parts/custom_sort_filter_proxy_model.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_sort_filter_proxy_model.h	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,16 @@
+#ifndef CUSTOM_SORT_FILTER_PROXY_MODEL_H
+#define CUSTOM_SORT_FILTER_PROXY_MODEL_H
+
+#include <QSortFilterProxyModel>
+#include <QModelIndex>
+
+class CustomSortFilterProxyModel : public QSortFilterProxyModel
+{
+public:
+	explicit CustomSortFilterProxyModel(QObject *parent = nullptr);
+protected:
+	bool filterAcceptsRow(int source_row, const QModelIndex& source_parent) const override;
+	bool lessThan(const QModelIndex &source_left, const QModelIndex &source_right) const override;
+};
+
+#endif // CUSTOM_SORT_FILTER_PROXY_MODEL_H
diff -ruN original_dirview/src/ui_parts/custom_system_model.cpp modified_dirview/src/ui_parts/custom_system_model.cpp
--- original_dirview/src/ui_parts/custom_system_model.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_system_model.cpp	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,142 @@
+#include <QFileInfo>
+#include <QPointer>
+#include <QDebug>
+
+#include "custom_system_model.h"
+#include "utils/format_utils.h"
+
+CustomSystemModel::CustomSystemModel(QThreadPool* threadPool, StopFlag* stopFlag, QObject* parent)
+	: QFileSystemModel(parent),
+	  m_thPool{threadPool ? threadPool : QThreadPool::globalInstance()},
+	  m_stopFlag{stopFlag ? stopFlag : nullptr}
+{}
+
+QVariant CustomSystemModel::data(const QModelIndex& index, int role) const
+{
+	if (!index.isValid() || index.column() != 1)
+	{
+		return QFileSystemModel::data(index, role);
+	}
+
+	switch (role)
+	{
+		case Qt::TextAlignmentRole:
+			return QVariant(Qt::AlignHCenter | Qt::AlignVCenter);
+
+		case CustomRoles::SizeInBytesRole:
+			return m_getRawSize(index);
+
+		case CustomRoles::InProgressRole:
+			return m_getInProgressStatus(index);
+
+		case Qt::DisplayRole:
+			return m_getFormattedSize(index, role);
+
+		default:
+			return QFileSystemModel::data(index, role);
+	}
+}
+
+QVariant CustomSystemModel::headerData(int section, Qt::Orientation orientation, int role) const
+{
+	if (role == Qt::TextAlignmentRole)
+	{
+		return QVariant(Qt::AlignHCenter | Qt::AlignVCenter);
+	}
+
+	return QFileSystemModel::headerData(section, orientation, role);
+}
+
+void CustomSystemModel::calculateAndSetDirSize(const QModelIndex& index)
+{
+	if (!index.isValid())
+	{
+		return;
+	}
+
+	QFileInfo info = fileInfo(index);
+
+	if (!info.isDir())
+	{
+		return;
+	}
+
+	QString path = filePath(index);
+
+	{
+		QWriteLocker lock(&m_cacheMutex);
+
+		if (m_inProgress.contains(path))
+		{
+			return;
+		}
+
+		m_inProgress.insert(path);
+	}
+
+	emit dataChanged(index, index, {CustomRoles::InProgressRole});
+
+	m_thPool->start([this, path, persistentIndex = QPersistentModelIndex(index), filters = filter()]() {
+		if (!persistentIndex.isValid())
+		{
+			m_removeFromInProgress(path);
+			return;
+		}
+
+		qint64 size = FileUtils::calculateDirSize(path, filters, m_stopFlag);
+
+		if (!persistentIndex.isValid() || (m_stopFlag && m_stopFlag->value()))
+		{
+			m_removeFromInProgress(path);
+			return;
+		}
+
+		{
+			QWriteLocker lock(&m_cacheMutex);
+			m_inProgress.remove(path);
+			m_sizeCache[path] = size;
+		}
+
+		emit dataChanged(persistentIndex, persistentIndex, {Qt::DisplayRole, CustomRoles::InProgressRole});
+	});
+}
+
+QVariant CustomSystemModel::m_getRawSize(const QModelIndex& index) const
+{
+	const auto fInfo = fileInfo(index);
+
+	if (fInfo.isDir())
+	{
+		QReadLocker lock(&m_cacheMutex);
+		return m_sizeCache.value(filePath(index), kDefaultDirSizeValue);
+	}
+	else
+	{
+		return fInfo.size();
+	}
+}
+
+QVariant CustomSystemModel::m_getInProgressStatus(const QModelIndex& index) const
+{
+	QReadLocker lock(&m_cacheMutex);
+	return m_inProgress.contains(filePath(index));
+}
+
+QVariant CustomSystemModel::m_getFormattedSize(const QModelIndex& index, int role) const
+{
+	qint64 sizeValue = kDefaultDirSizeValue;
+	{
+		QReadLocker lock(&m_cacheMutex);
+		sizeValue = m_sizeCache.value(filePath(index), kDefaultDirSizeValue);
+	}
+
+	return sizeValue != kDefaultDirSizeValue
+		? FormatUtils::formatSize(sizeValue)
+		: QFileSystemModel::data(index, role);
+}
+
+void CustomSystemModel::m_removeFromInProgress(const QString& path)
+{
+	QWriteLocker lock(&m_cacheMutex);
+	m_inProgress.remove(path);
+}
diff -ruN original_dirview/src/ui_parts/custom_system_model_delegate.cpp modified_dirview/src/ui_parts/custom_system_model_delegate.cpp
--- original_dirview/src/ui_parts/custom_system_model_delegate.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_system_model_delegate.cpp	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,82 @@
+// custom_system_model_delegate.cpp
+#include <QPainterPath>
+#include <QPen>
+#include <QStyleOptionButton>
+#include <QPushButton>
+#include <QStyle>
+
+#include "custom_system_model_delegate.h"
+
+CustomSystemModelDelegate::CustomSystemModelDelegate(QObject* parent)
+	: QStyledItemDelegate(parent)
+{}
+
+void CustomSystemModelDelegate::paint(
+	QPainter* painter,
+	const QStyleOptionViewItem& option,
+	const QModelIndex& index
+) const
+{
+	if (!index.isValid())
+	{
+		return;
+	}
+
+	const bool isInProgress = index.data(CustomRoles::InProgressRole).toBool();
+	const bool hasSize = !index.data(Qt::DisplayRole).toString().isEmpty();
+
+	// size already calculated and not in progress => use parent paint.
+	if (hasSize && !isInProgress)
+	{
+		QStyledItemDelegate::paint(painter, option, index);
+		return;
+	}
+
+	ButtonState state
+	{
+		static_cast<bool>(option.state & QStyle::State_MouseOver),
+		static_cast<bool>(option.state & QStyle::State_Selected),
+		isInProgress
+	};
+
+	m_paintButton(painter, option, state);
+}
+
+void CustomSystemModelDelegate::m_paintButton(
+	QPainter* painter,
+	const QStyleOptionViewItem& option,
+	const ButtonState& state
+) const
+{
+	static const QPushButton tempButton;
+
+	painter->save();
+	painter->setRenderHint(QPainter::Antialiasing);
+
+
+	QStyleOptionButton buttonOption;
+	buttonOption.initFrom(&tempButton);
+	buttonOption.rect = option.rect.adjusted(0, 0, 0, 0);
+	buttonOption.text = state.isInProgress
+		? kInProccessButtonText
+		: kCalculateButtonText;
+
+	if (state.isSelected || state.isHovered)
+	{
+		buttonOption.state |= QStyle::State_MouseOver;
+	}
+
+	if (state.isInProgress)
+	{
+		buttonOption.state |= QStyle::State_Sunken;
+	}
+
+	QApplication::style()->drawControl(
+		QStyle::CE_PushButton,
+		&buttonOption,
+		painter,
+		&tempButton
+	);
+
+	painter->restore();
+}
diff -ruN original_dirview/src/ui_parts/custom_system_model_delegate.h modified_dirview/src/ui_parts/custom_system_model_delegate.h
--- original_dirview/src/ui_parts/custom_system_model_delegate.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_system_model_delegate.h	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,36 @@
+// custom_system_model_delegate.h
+#ifndef CUSTOM_SYSTEM_MODEL_DELEGATE_H
+#define CUSTOM_SYSTEM_MODEL_DELEGATE_H
+
+#include <QStyledItemDelegate>
+#include <QModelIndex>
+#include <QPainter>
+#include <QColor>
+#include <QApplication>
+#include <QPalette>
+
+#include "utils/types.h"
+
+class CustomSystemModelDelegate : public QStyledItemDelegate
+{
+	Q_OBJECT
+public:
+	explicit CustomSystemModelDelegate(QObject* parent = nullptr);
+	void paint(
+		QPainter* painter,
+		const QStyleOptionViewItem& option,
+		const QModelIndex& index
+	) const override;
+
+private:
+	void m_paintButton(
+		QPainter* painter,
+		const QStyleOptionViewItem& option,
+		const ButtonState& state
+	) const;
+
+	static constexpr const char* kInProccessButtonText = "Processing";
+	static constexpr const char* kCalculateButtonText = "Calculate";
+};
+
+#endif // CUSTOM_SYSTEM_MODEL_DELEGATE_H
diff -ruN original_dirview/src/ui_parts/custom_system_model.h modified_dirview/src/ui_parts/custom_system_model.h
--- original_dirview/src/ui_parts/custom_system_model.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/src/ui_parts/custom_system_model.h	2026-06-28 02:38:03.088791803 +0300
@@ -0,0 +1,38 @@
+#ifndef CUSTOM_SYSTEM_MODEL_H
+#define CUSTOM_SYSTEM_MODEL_H
+
+#include <QFileSystemModel>
+#include <QVariant>
+#include <QModelIndex>
+#include <QThreadPool>
+#include <QReadWriteLock>
+#include <QHash>
+#include <QSet>
+
+#include "utils/file_utils.h"
+#include "utils/types.h"
+
+class CustomSystemModel : public QFileSystemModel
+{
+	Q_OBJECT
+public:
+	CustomSystemModel(QThreadPool* threadPool, StopFlag* _stopFlag, QObject* parent = nullptr);
+	QVariant data(const QModelIndex& index, int role = Qt::DisplayRole) const override;
+	QVariant headerData(int section, Qt::Orientation orientation, int role = Qt::DisplayRole) const override;
+	void calculateAndSetDirSize(const QModelIndex& index);
+private:
+	mutable QReadWriteLock m_cacheMutex;
+	QHash<QString, qint64> m_sizeCache;
+	QSet<QString> m_inProgress;
+	StopFlag* m_stopFlag;
+	QThreadPool* m_thPool;
+
+	QVariant m_getRawSize(const QModelIndex& index) const;
+	QVariant m_getInProgressStatus(const QModelIndex& index) const;
+	QVariant m_getFormattedSize(const QModelIndex& index, int role) const;
+	void m_removeFromInProgress(const QString& path);
+
+	static constexpr qint64 kDefaultDirSizeValue = -1;
+};
+
+#endif // CUSTOM_SYSTEM_MODEL_H
diff -ruN original_dirview/utils/CMakeLists.txt modified_dirview/utils/CMakeLists.txt
--- original_dirview/utils/CMakeLists.txt	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/CMakeLists.txt	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,15 @@
+add_library(utils STATIC
+	format_utils.h
+	format_utils.cpp
+	file_utils.h
+	file_utils.cpp
+	types.h
+)
+
+target_include_directories(utils PUBLIC
+	${CMAKE_CURRENT_SOURCE_DIR}/..
+)
+
+target_link_libraries(utils PUBLIC
+	Qt${QT_VERSION_MAJOR}::Core
+)
diff -ruN original_dirview/utils/file_utils.cpp modified_dirview/utils/file_utils.cpp
--- original_dirview/utils/file_utils.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/file_utils.cpp	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,57 @@
+#include <QVector>
+#include <QDirIterator>
+#include <QFileInfo>
+#include <QSet>
+
+#ifdef Q_OS_LINUX
+#include <sys/stat.h>
+#endif
+
+#include "file_utils.h"
+
+qint64 FileUtils::calculateDirSize(
+	const QString& path,
+	QDir::Filters filters,
+	StopFlag* stopFlag
+)
+{
+	qint64 totalSize = 0;
+	QDirIterator it(
+		path,
+		filters,
+		QDirIterator::Subdirectories
+	);
+
+	while (it.hasNext())
+	{
+		if (stopFlag && stopFlag->value())
+		{
+			break;
+		}
+
+		const QFileInfo info = it.next();
+
+		if (!info.isFile())
+		{
+			continue;
+		}
+
+		#ifdef Q_OS_LINUX
+			struct stat fileStat;
+
+			if (
+				stat(info.absoluteFilePath().toUtf8().constData(), &fileStat) == 0 &&
+				(fileStat.st_blocks <= 0)
+			)
+			{
+				continue;
+			}
+
+			totalSize += info.size();
+		#else
+			totalSize += info.size();
+		#endif
+	}
+
+	return totalSize;
+}
diff -ruN original_dirview/utils/file_utils.h modified_dirview/utils/file_utils.h
--- original_dirview/utils/file_utils.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/file_utils.h	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,20 @@
+#ifndef FILE_UTILS_H
+#define FILE_UTILS_H
+
+#include <QString>
+#include <QtGlobal>
+#include <QDir>
+
+#include "types.h"
+
+class FileUtils
+{
+public:
+	static qint64 calculateDirSize(
+		const QString& path,
+		QDir::Filters filters,
+		StopFlag* stopFlag = nullptr
+	);
+};
+
+#endif // FILE_UTILS_H
diff -ruN original_dirview/utils/format_utils.cpp modified_dirview/utils/format_utils.cpp
--- original_dirview/utils/format_utils.cpp	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/format_utils.cpp	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,21 @@
+#include <QLocale>
+
+#include "format_utils.h"
+
+QString FormatUtils::formatSize(qint64 bytes)
+{
+	static const QLocale locale;
+	int precision = 2;
+
+	if (bytes < kKb)
+	{
+		precision = 0;
+	}
+	else if (bytes < kMb)
+	{
+		precision = 1;
+	}
+
+
+	return locale.formattedDataSize(bytes, precision, QLocale::DataSizeIecFormat);
+}
diff -ruN original_dirview/utils/format_utils.h modified_dirview/utils/format_utils.h
--- original_dirview/utils/format_utils.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/format_utils.h	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,17 @@
+#ifndef FORMAT_UTILS_H
+#define FORMAT_UTILS_H
+
+#include <QString>
+#include <QLocale>
+#include <QtGlobal>
+
+class FormatUtils
+{
+public:
+	static QString formatSize(qint64 bytes);
+private:
+	static constexpr qint64 kKb = 1024;
+	static constexpr qint64 kMb = kKb * 1024;
+};
+
+#endif // FORMAT_UTILS_H
diff -ruN original_dirview/utils/types.h modified_dirview/utils/types.h
--- original_dirview/utils/types.h	1970-01-01 03:00:00.000000000 +0300
+++ modified_dirview/utils/types.h	2026-06-28 02:38:12.516792037 +0300
@@ -0,0 +1,41 @@
+#ifndef MY_TYPES
+#define MY_TYPES
+
+#include <atomic>
+#include <Qt>
+
+struct StopFlag
+{
+public:
+	inline StopFlag()
+	{
+		setFlag(false);
+	}
+
+	inline void setFlag(bool flag)
+	{
+		m_flag.store(flag);
+	}
+
+	inline bool value() const
+	{
+		return m_flag.load();
+	}
+private:
+	std::atomic<bool> m_flag;
+};
+
+struct ButtonState
+{
+	bool isHovered;
+	bool isSelected;
+	bool isInProgress;
+};
+
+enum CustomRoles
+{
+	InProgressRole = Qt::UserRole + 1,
+	SizeInBytesRole = Qt::UserRole + 2
+};
+
+#endif // MY_TYPES
